docs: Update VM support to use cilium-cli (#15320)
cilium-team/cilium: Documentation/gettingstarted/external-workloads.rst

  in the :ref:`firewall_requirements` section.

* This guide assumes your external workload manages domain name
  resolution service by a stand-alone ``/etc/resolv.conf``, or via
  systemd (e.g., Ubuntu).

* So far this functionality is only tested with the vxlan tunneling
  datapath mode (default for most installations).

Prepare your cluster
####################

Enable support for external workloads
=====================================

Your cluster must be configured with support for external workloads
enabled. This can be done with the cilium CLI tool by issuing ``cilium
clustermesh enable`` after ``cilium install``:

.. parsed-literal::

>     cilium install --config tunnel=vxlan

Reviewer: Nit: The directive above this block no longer needs to be a `.. parsed-literal::`, could you please switch to `.. code-block:: shell-session`?

Author: Thanks for the guidance; changed all directives on shell commands/output to `.. code-block:: shell-session`

    otherwise becomes unreachable.

This will add a deployment for ``clustermesh-apiserver`` into your
cluster, as well as the related cluster resources, such as TLS
secrets.

``clustermesh-apiserver`` service is exposed to the external
workloads. If your are on GKE, EKS, or AKS, this is done using the
``LoadBalancer`` service type that allows use of a static
IP.

Tell your cluster about external workloads
==========================================

To allow an external workload to join your cluster, the cluster must
be informed about each such workload. This is done by creating a
``CiliumExternalWorkload`` (CEW) resource for each external
workload. CEW resource specifies the name and identity labels
(including namespace) for the workload. The name must be the hostname
of the external workload, as returned by the ``hostname`` command run
in the external workload. In this example this is ``runtime``. For now
you must also allocate a small IP CIDR that must be unique to each
workload. For example, for a VM named ``runtime`` that is to join the
``default`` namespace (``vm`` is an alias for the
``external-workload`` subcommand):

.. parsed-literal::

>     cilium clustermesh vm create runtime -n default --ipv4-alloc-cidr 10.192.1.0/30

Reviewer: Same remark on the `.. parsed-literal::`

``-n`` is an alias for ``--namespace`` and can be left out when the
value is ``default``. The namespace value will be set as an identity
label. The CEW resource itself is not namespaced.

To see the list of existing CEW resources, run:

> .. parsed-literal::

Reviewer: ```suggestion
.. code-block:: shell-session
```

    cilium clustermesh vm status

Note that CEW resources are not namespaced, so this command shows the
status of all CEW resources regardless of the namespace label that was
used when creating them. ``--namespace`` option for the status command
controls the namespace of Cilium deployment in your cluster and
usually needs to be left as the default ``kube-system``.

At this point the ``IP:`` in the status for ``runtime`` is ``N/A`` to
inform that the VM has not yet joined the cluster.

Install and configure Cilium on external workloads
##################################################

Run the external workload install command on your k8s cluster. This

> extracts that TLS certificates and other access information from the

Reviewer: ```suggestion
extracts the TLS certificates and other access information from the
```

Author: Fixed, thanks!

cluster installation and writes out an installation script to be used
in the external workloads to install Cilium and connect it to your k8s
cluster:

> .. parsed-literal::

Reviewer: ```suggestion
.. code-block:: shell-session
```

* This guide assumes your external workload manages domain name
  resolution service via systemd.

* So far this functionality is only tested with the vxlan tunneling
  datapath mode (default for most installations).

> 

Reviewer 2: Initially, I tested this guide on the runtime VM which still has 4.9 kernel and thus doesn't support host reachable services which the Cilium agent started by the `install-external-workload.sh` script enables. Might be worth to point out in the prerequisites that external workloads thus require at least kernel 4.17 where BPF host reachable services are supported? Or do we not strictly needs host reachable services, in which case it might be good to remove `--enable-host-reachable-services` from the generated install script.

Author: Unfortunately ``--enable-host-reachable-services`` is a strict requirement. This is documented in the *Prerequisites* section of the GSG.

Reviewer 2: Sorry, I somehow missed that during the review. It's completely fine as is.

Prepare your cluster
####################

Enable support for external workloads
=====================================

Your cluster must be configured with support for external workloads

> enabled. This can be with the cilium CLI tool by issuing ``cilium

Reviewer 2: Nit:
```suggestion
enabled. This can be done with the cilium CLI tool by issuing ``cilium
```

Author: Fixed.

clustermesh enable`` after ``cilium install``:

.. parsed-literal::

    cilium install

>     cilium clustermesh enable

Reviewer 2: When testing this with a GKE cluster, I had to provide the `--service-type` (as described below) explicitly. As far as I remember, this wasn't the case without cilium/cilium-cli#123, i.e. with the latest `master` build of the Cilium CLI. I'm not sure whether this just assumed NodePort as default or whether the auto-detection for GKE somehow got broken by cilium/cilium-cli#123.

I first used `--service-type NodePort` as suggested, but this contradicts the statement below:

> ``clustermesh-apiserver`` service is exposed to the external workloads. If your are on GKE, EKS, or AKS, this is done using the TLS configuration ``LoadBalancer`` service type that allows use of a static IP.

This might be a bit confusing (at least it was to me when initially reading this).

On my second attempt I used `--service-type LoadBalancer` and everything worked fine, so maybe the note should suggest LoadBalancer explicitly for managed k8s environments?

Author: The behavior should not have changed here, I'll push a commit to `cilium-cli` fix this up!

Reviewer 2: Thanks!

.. note::

>     If this fails indicating that --service-type needs to be given,

Reviewer 2: ```suggestion
    If this fails indicating that ``--service-type`` needs to be given,
```

Author: Fixed.

>     add ``--service-type NodePort`` to the second command above. This

Reviewer 2: ```suggestion
    add ``--service-type NodePort`` to the second command above, i.e. ``cilium lustermesh enable --service-type NodePort``. This
```

Author: Added.

>     will allow you to go though this guide, but be warned that

Reviewer 2: ```suggestion
    will allow you to go through this guide, but be warned that
```

Author: Fixed.

    NodePort service type makes your installation very fragile, it

>     will become non-functional if the node though which the service is

Reviewer 2: ```suggestion
    will become non-functional if the node through which the service is
```

Author: Fixed.

    accessed is removed from the cluster or if it otherwise becomes
    unreachable.

This will add a deployment for ``clustermesh-apiserver`` into your
cluster, as well as the related cluster resources, such as TLS
secrets.

``clustermesh-apiserver`` service is exposed to the external
workloads. If your are on GKE, EKS, or AKS, this is done using the
``LoadBalancer`` service type that allows use of a static
IP.

Tell your cluster about external workloads
==========================================

To allow an external workload to join your cluster, the cluster must
be informed about each such workload. This is done by creating a
``CiliumExternalWorkload`` (CEW) resource for each external
workload. CEW resource specifies the name and identity labels (including
namespace) for the workload. For now you must also allocate a small IP
CIDR that must be unique to each workload. For example, for a VM named
``runtime`` that is to join the ``default`` namespace (``vm`` is an
alias for the ``external-workload`` subcommand):

.. parsed-literal::

    cilium clustermesh vm create runtime -n default --ipv4-alloc-cidr 10.192.1.0/30

> '-n' is an alias for '--namespace' and can be left out when the value

Reviewer 2: ```suggestion
``-n`` is an alias for ``--namespace`` and can be left out when the value
```

Author: Fixed.

> is ``default``. Namespace value will be set as an identity label, CEW

Reviewer 2: ```suggestion
is ``default``. The namespace value will be set as an identity label. The CEW
```

Author: Fixed.

    hostname

Next, copy ``install-external-workload.sh`` created above to the
external workload. Then run the installation script:

.. parsed-literal::

>     ./install-external-workload.sh

Reviewer: Same remark on the `.. parsed-literal::`.

.. parsed-literal::

>     KVStore:     Ok   etcd: 1/1 connected, lease-ID=7c02748328e75f57, lock lease-ID=7c02748328e75f59, has-quorum=true: https://clustermesh-apiserver.cilium.io:32379 - 3.4.13 (Leader)

Reviewer: Same remark on the `.. parsed-literal::`.

    Kubernetes:  Disabled
    ...

Check that cluster DNS works:

.. code:: bash

>     nslookup -norecurse clustermesh-apiserver.kube-system.svc.cluster.local

Reviewer: Same remark for the `.. code:: bash`, should also be a  code block for shell session.

        - matchLabels:
            io.kubernetes.pod.name: runtime

The ping should continue if you delete the policy:

> .. parsed-literal::

Reviewer: ```suggestion
.. code-block:: shell-session
```

     kubectl delete ccnp test-ccnp

Clean-up
########

You can remove the Cilium installation from your external workload by
running the installation script with the ``uninstall`` argument:

> .. parsed-literal::

Reviewer: ```suggestion
.. code-block:: shell-session
```
